src/nodes.coffee

```diff
@@ -2289,93 +2268,6 @@ exports.Assign = class Assign extends Base
-
-    # Recursive function for searching and storing rest elements in objects.
-    # e.g. `{[properties...]} = source`.
-    traverseRest = (properties, source) =>
-      restElements = []
-      restIndex = undefined
-      source = new Value source unless source.properties?
-
-      for prop, index in properties
-        nestedSourceDefault = nestedSource = nestedProperties = null
-        if prop instanceof Assign
-          # prop is `k: expr`, we need to check `expr` for nested splats
-          if prop.value.isObject?()
-            # prop is `k = {...} `
-            continue unless prop.context is 'object'
-            # prop is `k: {...}`
-            nestedProperties = prop.value.base.properties
-          else if prop.value instanceof Assign and prop.value.variable.isObject()
-            # prop is `k: {...} = default`
-            nestedProperties = prop.value.variable.base.properties
-            [prop.value.value, nestedSourceDefault] = prop.value.value.cache o
-          if nestedProperties
-            nestedSource = new Value source.base, source.properties.concat [new Access getPropKey prop]
-            nestedSource = new Value new Op '?', nestedSource, nestedSourceDefault if nestedSourceDefault
-            restElements.push traverseRest(nestedProperties, nestedSource)...
-        else if prop instanceof Splat
-          prop.error "multiple rest elements are disallowed in object destructuring" if restIndex?
-          restIndex = index
-          restElements.push {
-            name: prop.name.unwrapAll()
-            source
-            excludeProps: new Arr (getPropName p for p in properties when p isnt prop)
-          }
-
-      if restIndex?
-        # Remove rest element from the properties after iteration
-        properties.splice restIndex, 1
-
-      restElements
-
-    # Cache the value for reuse with rest elements.
-    valueRefTemp =
-      if @value.shouldCache()
-        new IdentifierLiteral o.scope.freeVariable 'ref', reserve: false
-      else
-        @value.base
-
-    # Find all rest elements.
-    restElements = traverseRest @variable.base.properties, valueRefTemp
-    return no unless restElements and restElements.length > 0
-
-    [@value, valueRef] = @value.cache o
-    result = new Block [@]
-
-    for restElement in restElements
-      value = new Call new Value(new Literal utility 'objectWithoutKeys', o), [restElement.source, restElement.excludeProps]
-      result.push new Assign new Value(restElement.name), value, null, param: if @param then 'alwaysDeclare' else null
-
-    fragments = result.compileToFragments o
-    if o.level is LEVEL_TOP
-      # Remove leading tab and trailing semicolon
-      fragments.shift()
-      fragments.pop()
-
-    fragments
-
   # Brief implementation of recursive pattern matching, when assigning array or
   # object literals to a value. Peeks at their properties to assign inner names.
   compileDestructuring: (o) ->
```

Comment:
@zdenko I believe you can also stop treating object spreads as `complexObjects()` in `compileDestructuring()`? So eg `[..., {r...}] = arr` compiles to
```
[{...r}] = slice.call(arr, -1);
```
instead of
```
ref = slice.call(arr, -1), ({...r} = ref[0]);
```
If so, not sure exactly what code inside `compileDestructuring()` can be removed, at least the stuff related to `hasObjSpreads()`?